File: src/0819.most-common-word.837/solution.ts
function mostCommonWord (paragraph: string, banned: string[]): string {
    paragraph = paragraph.replace(/[!?',;.]/g, ' ').toLowerCase();
    const words = paragraph.split(' ');
    const bannedSet = new Set(banned);
    const map = new Map<string, number>();
    for (const w of words) {
        if (w === ' ' || w === '' || bannedSet.has(w)) {
            continue;
        }
        map.set(w, map.has(w) ? map.get(w) + 1 : 1);
    }

    let max = 0;
    let result = '';
    for (const [w, c, ] of map) {
        if (c > max) {
            max = c;
            result = w;
        }
    }

    return result;
}

File: src/0819.most-common-word.837/solution.java
class Solution {
    private final Map<String,Integer> map = new HashMap<>();

    public String mostCommonWord(String paragraph, String[] banned) {
        Set<String> set = new HashSet<>(Arrays.asList(banned));

        char[] chars = paragraph.toLowerCase().toCharArray();
        StringBuilder stringBuilder = new StringBuilder();
        int index = 0;
        while (index<chars.length){
            char c = chars[index++];
            if(c == ' ' || c == '!' || c == '?' || c == '\'' || c == ',' || c == ';' || c == '.'){
                updateCount(stringBuilder);
                continue;
            }
            stringBuilder.append(c);
        }
        updateCount(stringBuilder);
        String result = "";
        int maxCount = 0;
        for(Map.Entry<String,Integer> entry : map.entrySet() ){
            if(set.contains(entry.getKey())){
                continue;
            }
            if(entry.getValue()>maxCount){
                result = entry.getKey();
                maxCount = entry.getValue();
            }
        }

        return result;
    }


    private void updateCount(StringBuilder s){
        if(!s.isEmpty()){
            map.merge(s.toString(),1, Integer::sum);
            s.delete(0,s.length());
        }


    }
}